### PR TITLE
6 3 starting an application form not completed

### PR DESCRIPTION
**Pull request by pcbennett108**
This PR Complete's user story 3,

Added validation rules to Application controller. Since these are mailing addresses I stuck with the post office conventions of 2 letter states and 5 digit zip codes. (I am disregarding the -4 zip codes to keep it simple).

Updated the Application seed file entries to conform to new validations.

The flash:alert method used will list out all fields that do not validate.

### CODE REVIEW

File: app/models/application.rb [hunk 1]
```diff
@@ -1,4 +1,8 @@
 class Application < ApplicationRecord
   has_many :application_pets
   has_many :pets, through: :application_pets
+
+  validates :name, :street, :city, :state, :zip, :description, presence: true
+  validates :zip, length: { is: 5 }
+  validates :state, format: { with: /[A-Z]{2}/, message: "Please use 2 capital letters for the state abbreviation" }
```
Comment (ethanrossblack) [inline app/models/application.rb:7]:
```suggestion
  validates :state, format: { with: /[A-Z]{2}/, message: "must be an abbreviation of two capital letters" }
```

File: spec/features/applications/create_spec.rb [hunk 2]
```diff
@@ -63,5 +63,41 @@
       expect(page).to have_content("In Progress")
     end
 
+    it "if form is not filled in, will display an error and re-render the form" do
+      visit "/applications/new"
+      click_on "Submit"
+
+      expect(page).to have_current_path("/applications/new")
+      expect(page).to have_content("Error: Name can't be blank, Street can't be blank, City can't be blank, State can't be blank, Zip can't be blank, Description can't be blank, Zip is the wrong length (should be 5 characters), State Please use 2 capital letters for the state abbreviation")
```
Comment (ethanrossblack) [inline spec/features/applications/create_spec.rb:71]:
```suggestion
      expect(page).to have_content("Error: Name can't be blank, Street can't be blank, City can't be blank, State can't be blank, Zip can't be blank, Description can't be blank, Zip is the wrong length (should be 5 characters), State must be an abbreviation of two capital letters")
```

File: app/models/application.rb (hunk 1 above)
Comment (ethanrossblack) [reply app/models/application.rb:7]: The grammar of `State Please use 2 capital letters for the state abbreviation` comes across as unnatural. I changed the error message up a little bit to make it flow better (also put this in the test validation).

I think ultimately, we could refactor state to be a dropdown of all 50 states (still only working with US pet adoption).

File: spec/features/applications/create_spec.rb [hunk 3]
```diff
@@ -2,8 +2,8 @@
 
 RSpec.describe "application creation" do
   before(:each) do
-    @mr_ape = Application.create!(name: "Mr. Ape", street: "123 Turing Lane", city: "Boulder", state: "Colorado", zip: "80301", description: "I really want a dog because I love dogs", status: "In Progress")
-    @penny_lane = Application.create!(name: "Penny Lane", street: "555 McCartney", city: "Hollywood", state: "California", zip: "90210", description: "Strawberry Fields Forever", status: "In Progress")
+    @mr_ape = Application.create!(name: "Mr. Ape", street: "123 Turing Lane", city: "Boulder", state: "CO", zip: "80301", description: "I really want a dog because I love dogs", status: "In Progress")
+    @penny_lane = Application.create!(name: "Penny Lane", street: "555 McCartney", city: "Hollywood", state: "CA", zip: "90210", description: "Strawberry Fields Forever", status: "In Progress")
```
Comment (ethanrossblack) [inline spec/features/applications/create_spec.rb:6]: If we change the state length here, we'll need to change it in every other test sample data, too.

File: spec/features/applications/create_spec.rb [hunk 4]
```diff
@@ -63,5 +63,41 @@
...
+      fill_in "Name", with: "Jill"
+      fill_in "Street", with: "99 Memory Ln."
+      fill_in "city", with: "Townsville"
+      fill_in "State", with: "az"
+      fill_in "Zip", with: "1234"
+      fill_in "Description", with: "The liger is my fav animal"
+      click_on "Submit"
+
+      expect(page).to have_current_path("/applications/new")
+      expect(page).to have_content("Error: Zip is the wrong length (should be 5 characters), State Please use 2 capital letters for the state abbreviation")
```
Comment (ethanrossblack) [inline spec/features/applications/create_spec.rb:100]:
```suggestion
      expect(page).to have_content("Error: Zip is the wrong length (should be 5 characters), State must be an abbreviation of two capital letters")
```

Actually, I think we should be able to accept a state input (if we do it by typing) if it's lowercase. It's very easy to fix this with presentation logic, or in a controller.

**Review by ethanrossblack — CHANGES_REQUESTED** [review]
See comments

**Review by ethanrossblack — APPROVED** [review]
I updated the test data in `spec/features/applications/index_spec.rb` and `spec/features/applications/show_spec.rb` to reflect new flash errors and validations. Looks good now!

File: app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -12,7 +12,7 @@
 
   <body>
   <% flash.each do |type, msg| %>
-    <div>
+    <div style="color:red">
```
[the remark this answers is not in the document]
Comment (pcbennett108) [reply app/views/layouts/application.html.erb]: I made the change per your suggestion. My initial logic was that it's such a small thing to turn one error message red, I'll just do it here. But you're right about there being a preferred way to style.

File: db/seeds.rb
[the remark this answers is not in the document]
Comment (pcbennett108) [reply db/seeds.rb:1]: Pulled updates and fixed conflict.

File: db/seeds.rb [hunk 6]
```diff
@@ -34,10 +34,10 @@
...
-paul = Application.create(name: "Paul", street: "1960 Penny Lane", city: "Bedfordshire", state: "England", zip: "48", description: "I still believe love is all you need.  I don't know a better message than that.", status: "Accepted")
-ringo = Application.create(name: "Ringo", street: "235 River Ferry", city: "Liverpool", state: "England", zip: "45", description: "I get by with a little help from my friends", status: "Rejected")
+mr_ape = Application.create(name: "Mr. Ape", street: "123 Turing Lane", city: "Boulder", state: "CO", zip: "80301", description: "I really want a dog because I love dogs", status: "In Progress")
+penny_lane = Application.create(name: "Penny Lane", street: "555 McCartney", city: "Hollywood", state: "CA", zip: "90210", description: "Strawberry Fields Forever", status: "Pending")
+paul = Application.create(name: "Paul", street: "1960 Penny Lane", city: "Bedfordshire", state: "UK", zip: "48J77", description: "I still believe love is all you need.  I don't know a better message than that.", status: "Accepted")
```
Comment (garussell) [inline db/seeds.rb:39]: I see what you were referring to with formatting these zipcodes.

**Review by garussell — APPROVED** [review]
Looks good I approve.

**Pull request merged**